# Installation {#installation .unnumbered}

This section describes different ways to install `pyANI-plus` on the most common operating systems such as Unix/Linux and macOS. 

Currently, we support three ways to install `pyANI-plus` on your system:

1. Installation from source (i.e. download/clone from GitHub)
2. Installation with Miniconda/Anaconda/Bioconda
2. Installation via `pip`/PyPI


## Installation from source

To install `pyANI-plus` from source, you can either download it from the [Releases page](https://github.com/pyani-plus/pyani-plus/releases/) or clone the repository using `git`.

To get the latest version with `git`, run the following command in a terminal:

```bash
git clone https://github.com/pyani-plus/pyani-plus
```

Alternatively you can visit the [Release](https://github.com/pyani-plus/pyani-plus/releases) page and click on one of the avaliable versions to get the source code in compressed form. 

Once the download is complete (and uncompressed, if required), navigate to the repository:

```bash
cd pyani-plus
```

Then install `pyANI-plus` by running the appropriate script for your operating system:

```bash
make install_linux # For Linux
make install_macos # For macOS
```

## Installation with Miniconda/Anaconda/Bioconda

With a working Miniconda/Anacodna installation that has Bioconda available, install using the command

```bash
conda install pyani-plus
```

## Installation via `pip`/PyPI

`Pyani-plus` can be installed using the command

```bash
pip install pyani-plus
```

## Check the installation

To check if the installation was sucessful, run the following command:

```bash
pyani-plus --help
```

If the installation was completed correctly, this should display a list of available commands and options, similar to this:

```bash
% pyani-plus -h

 Usage: pyani-plus [OPTIONS] COMMAND [ARGS]...

╭─ Options ──────────────────────────────────────────────────────────────────────────────╮
│ --version             -v        Show tool version (on stdout) and quit.                │
│ --install-completion            Install completion for the current shell.              │
│ --show-completion               Show completion for the current shell, to copy it or   │
│                                 customize the installation.                            │
│ --help                -h        Show this message and exit.                            │
╰────────────────────────────────────────────────────────────────────────────────────────╯
╭─ Commands ─────────────────────────────────────────────────────────────────────────────╮
│ resume               Resume any (partial) run already logged in the database.          │
│ list-runs            List the runs defined in a given pyANI-plus SQLite3 database.     │
│ delete-run           Delete any single run from the given pyANI-plus SQLite3 database. │
│ export-run           Export any single run from the given pyANI-plus SQLite3 database. │
│ plot-run             Plot heatmaps and distributions for any single run.               │
│ plot-run-comp        Plot comparisons between multiple runs.                           │
│ classify             Classify genomes into clusters based on ANI results.              │
╰────────────────────────────────────────────────────────────────────────────────────────╯
╭─ ANI methods ──────────────────────────────────────────────────────────────────────────╮
│ anim                 Execute ANIm calculations, logged to a pyANI-plus SQLite3         │
│                      database.                                                         │
│ dnadiff              Execute mumer-based dnadiff calculations, logged to a pyANI-plus  │
│                      SQLite3 database.                                                 │
│ anib                 Execute ANIb calculations, logged to a pyANI-plus SQLite3         │
│                      database.                                                         │
│ fastani              Execute fastANI calculations, logged to a pyANI-plus SQLite3      │
│                      database.                                                         │
│ sourmash             Execute sourmash-plugin-branchwater ANI calculations, logged to a │
│                      pyANI-plus SQLite3 database.                                      │
│ external-alignment   Compute pairwise ANI from given multiple-sequence-alignment (MSA) │
│                      file.                                                             │
╰────────────────────────────────────────────────────────────────────────────────────────╯
```

{{< include includes/_requirements.qmd >}}
