<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.43">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Leighton Pritchard">

<title>pyANI-plus walkthrough – pyANI-plus</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./subcommands/subcommands.html" rel="next">
<link href="./installation.html" rel="prev">
<link href="./icon_32.png" rel="icon" type="image/png">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-d4d76bf8491c20bad77d141916dc28e1.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-a34d670291f06f286357e447776a572a.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script src="site_libs/quarto-contrib/glightbox/glightbox.min.js"></script>
<link href="site_libs/quarto-contrib/glightbox/glightbox.min.css" rel="stylesheet">
<link href="site_libs/quarto-contrib/glightbox/lightbox.css" rel="stylesheet">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body class="nav-sidebar floating slimcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./walkthrough.html">pyANI-plus walkthrough</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">pyANI-plus</a> 
        <div class="sidebar-tools-main">
    <a href="https://github.com/sipbs-compbiol/REPOSITORY" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
  <a href="" class="quarto-reader-toggle quarto-navigation-tool px-1" onclick="window.quartoToggleReader(); return false;" title="Toggle reader mode">
  <div class="quarto-reader-toggle-btn">
  <i class="bi"></i>
  </div>
</a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">About pyANI-plus</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./requirements.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Requirements</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./installation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Installation</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./walkthrough.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">pyANI-plus walkthrough</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./subcommands/subcommands.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">pyANI-plus subcommands</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./subcommands/anib.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title"><code>anib</code></span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./subcommands/anim.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title"><code>anim</code></span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./subcommands/dnadiff.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title"><code>dnadiff</code></span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./subcommands/external_alignment.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title"><code>external-alignment</code></span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./subcommands/fastani.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title"><code>fastani</code></span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./subcommands/sourmash.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title"><code>sourmash</code></span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./subcommands/plot_run.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title"><code>plot-run</code></span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./subcommands/plot_run_comp.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title"><code>plot-run-comp</code></span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./subcommands/list_runs.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title"><code>list-runs</code></span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./subcommands/export_run.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title"><code>export-run</code></span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./subcommands/resume.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title"><code>resume</code></span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./subcommands/delete_run.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title"><code>delete-run</code></span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./subcommands/classify.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title"><code>classify</code></span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./testing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Testing</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a href="./licensing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Licensing</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./pyani_licence.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">pyANI-plus</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./doc_licence.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">pyANI-plus documentation</span></span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#collect-genomes-for-analysis" id="toc-collect-genomes-for-analysis" class="nav-link active" data-scroll-target="#collect-genomes-for-analysis">Collect genomes for analysis</a></li>
  <li><a href="#conducting-ani-analysis" id="toc-conducting-ani-analysis" class="nav-link" data-scroll-target="#conducting-ani-analysis">Conducting ANI analysis</a>
  <ul class="collapse">
  <li><a href="#conduct-anib-analysis" id="toc-conduct-anib-analysis" class="nav-link" data-scroll-target="#conduct-anib-analysis">Conduct ANIb analysis</a></li>
  <li><a href="#conduct-anim-analysis" id="toc-conduct-anim-analysis" class="nav-link" data-scroll-target="#conduct-anim-analysis">Conduct ANIm analysis</a></li>
  <li><a href="#conduct-dnadiff-analysis" id="toc-conduct-dnadiff-analysis" class="nav-link" data-scroll-target="#conduct-dnadiff-analysis">Conduct dnadiff analysis</a></li>
  <li><a href="#conduct-fastani-analysis" id="toc-conduct-fastani-analysis" class="nav-link" data-scroll-target="#conduct-fastani-analysis">Conduct fastani analysis</a></li>
  <li><a href="#conduct-sourmash-analysis" id="toc-conduct-sourmash-analysis" class="nav-link" data-scroll-target="#conduct-sourmash-analysis">Conduct sourmash analysis</a></li>
  </ul></li>
  <li><a href="#reporting-analyses-and-analysis-results" id="toc-reporting-analyses-and-analysis-results" class="nav-link" data-scroll-target="#reporting-analyses-and-analysis-results">Reporting Analyses and Analysis Results</a>
  <ul class="collapse">
  <li><a href="#list-runs" id="toc-list-runs" class="nav-link" data-scroll-target="#list-runs">List all runs in the database</a></li>
  <li><a href="#exporting-ani-results-in-a-tabular-format" id="toc-exporting-ani-results-in-a-tabular-format" class="nav-link" data-scroll-target="#exporting-ani-results-in-a-tabular-format">Exporting ANI results in a tabular format</a></li>
  <li><a href="#graphical-output" id="toc-graphical-output" class="nav-link" data-scroll-target="#graphical-output">Graphical output</a></li>
  </ul></li>
  <li><a href="#plotting-comparisons-between-runs" id="toc-plotting-comparisons-between-runs" class="nav-link" data-scroll-target="#plotting-comparisons-between-runs">Plotting comparisons between runs</a></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/sipbs-compbiol/REPOSITORY/edit/main/walkthrough.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/sipbs-compbiol/REPOSITORY/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content page-columns page-full" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">pyANI-plus walkthrough</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>This section walks you through how <code>pyANI-plus</code> can be applied to calculate Average Nucleotide Identity, render graphical and tabular summary output, and perform other related measures for whole genome comparisons. The general procedue for any <code>pyANI-plus</code> analysis is:</p>
<ol type="1">
<li>Collect genomes for analysis</li>
<li>Perform ANI analysis using diffrent methods such as ANIb, ANIm etc.</li>
<li>Report and visualise analysis report</li>
<li>Use the analysis results to classify input genomes and generate species hypotheses</li>
</ol>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>Before using <code>pyANI-plus</code>, make sure to install it on a local machine like a laptop, desktop, server, or cluster. Please see section <a href="./installation.html#installation">installation</a> for installation instructions.</p>
</div>
</div>
<p>This is a command-line tool, meaning you type commands into a terminal window to run it. To view the avaliable options we type <code>pyani-plus</code> (in lower case), space, then <code>-h</code> (minus lower-case H) for the help option, and finally enter or return to run the command:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">pyani-plus</span> <span class="at">-h</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>This should output the following - hopefully in colour depending on your terminal setup:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a> <span class="ex">Usage:</span> pyani-plus <span class="pp">[</span><span class="ss">OPTIONS</span><span class="pp">]</span> COMMAND <span class="pp">[</span><span class="ss">ARGS</span><span class="pp">]</span>...</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="ex">╭─</span> Options ────────────────────────────────────────────────────────────────╮</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span> <span class="at">--version</span>             <span class="at">-v</span>        Show tool version <span class="er">(</span><span class="ex">on</span> stdout<span class="kw">)</span> <span class="ex">and</span> quit.  │</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span> <span class="at">--install-completion</span>            Install completion for the current       │</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>                                 shell.                                   │</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span> <span class="at">--show-completion</span>               Show completion for the current shell,   │</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>                                 to copy it or customize the              │</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>                                 installation.                            │</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span> <span class="at">--help</span>                <span class="at">-h</span>        Show this message and exit.              │</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a><span class="ex">╰──────────────────────────────────────────────────────────────────────────╯</span></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a><span class="ex">╭─</span> Commands ───────────────────────────────────────────────────────────────╮</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span> resume               Resume any <span class="er">(</span><span class="ex">partial</span><span class="kw">)</span> <span class="ex">run</span> already logged in the      │</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>                      database.                                           │</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span> list-runs            List the runs defined in a given pyANI-plus SQLite3 │</span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>                      database.                                           │</span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span> delete-run           Delete any single run from the given pyANI-plus     │</span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>                      SQLite3 database.                                   │</span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span> export-run           Export any single run from the given pyANI-plus     │</span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>                      SQLite3 database.                                   │</span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span> plot-run             Plot heatmaps and distributions for any single run. │</span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span> plot-run-comp        Plot comparisons between multiple runs.             │</span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span> classify             Classify genomes into clusters based on ANI         │</span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>                      results.                                            │</span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a><span class="ex">╰──────────────────────────────────────────────────────────────────────────╯</span></span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true" tabindex="-1"></a><span class="ex">╭─</span> ANI methods ────────────────────────────────────────────────────────────╮</span>
<span id="cb2-27"><a href="#cb2-27" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span> anim                 Execute ANIm calculations, logged to a pyANI-plus   │</span>
<span id="cb2-28"><a href="#cb2-28" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>                      SQLite3 database.                                   │</span>
<span id="cb2-29"><a href="#cb2-29" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span> dnadiff              Execute mumer-based dnadiff calculations, logged to │</span>
<span id="cb2-30"><a href="#cb2-30" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>                      a pyANI-plus SQLite3 database.                      │</span>
<span id="cb2-31"><a href="#cb2-31" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span> anib                 Execute ANIb calculations, logged to a pyANI-plus   │</span>
<span id="cb2-32"><a href="#cb2-32" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>                      SQLite3 database.                                   │</span>
<span id="cb2-33"><a href="#cb2-33" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span> fastani              Execute fastANI calculations, logged to a           │</span>
<span id="cb2-34"><a href="#cb2-34" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>                      pyANI-plus SQLite3 database.                        │</span>
<span id="cb2-35"><a href="#cb2-35" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span> sourmash             Execute sourmash-plugin-branchwater ANI             │</span>
<span id="cb2-36"><a href="#cb2-36" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>                      calculations, logged to a pyANI-plus SQLite3        │</span>
<span id="cb2-37"><a href="#cb2-37" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>                      database.                                           │</span>
<span id="cb2-38"><a href="#cb2-38" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span> external-alignment   Compute pairwise ANI from given                     │</span>
<span id="cb2-39"><a href="#cb2-39" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>                      multiple-sequence-alignment <span class="er">(</span><span class="ex">MSA</span><span class="kw">)</span> <span class="ex">file.</span>             │</span>
<span id="cb2-40"><a href="#cb2-40" aria-hidden="true" tabindex="-1"></a><span class="ex">╰──────────────────────────────────────────────────────────────────────────╯</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>To see the options for a specific subcommand, use <code>pyani-plus &lt;subcommand&gt; -h</code>. For example, to view options for the ANIb method:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">pyani-plus</span> anib <span class="at">-h</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Expected output:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a> <span class="ex">Usage:</span> pyani-plus anib <span class="pp">[</span><span class="ss">OPTIONS</span><span class="pp">]</span> FASTA</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a> <span class="ex">Execute</span> ANIb calculations, logged to a pyANI-plus SQLite3 database.</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="ex">╭─</span> Arguments ──────────────────────────────────────────────────────────────╮</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span> <span class="pp">*</span>    fasta      PATH  Directory of FASTA files <span class="er">(</span><span class="ex">extensions</span> .fas, .fasta, │</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>                       .fna<span class="kw">)</span><span class="bu">.</span>                                             │</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>                       <span class="pp">[</span><span class="ss">required</span><span class="pp">]</span>                                         │</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a><span class="ex">╰──────────────────────────────────────────────────────────────────────────╯</span></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a><span class="ex">╭─</span> Options ────────────────────────────────────────────────────────────────╮</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span> <span class="pp">*</span>  <span class="at">--database</span>   <span class="at">-d</span>      FILE           Path to pyANI-plus SQLite3        │</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>                                        database.                         │</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>                                        <span class="pp">[</span><span class="ss">required</span><span class="pp">]</span>                        │</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>    <span class="at">--name</span>               TEXT           Run name. Default is <span class="st">'N genomes   │</span></span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a><span class="st">│                                        using METHOD'</span>.                    │</span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>    <span class="at">--create-db</span>                         Create database if does not       │</span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>                                        exist.                            │</span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>    <span class="at">--executor</span>           [local<span class="kw">|</span><span class="ex">slurm]</span>  How should the internal tools be  │</span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>                                        run<span class="pp">?</span>                              │</span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>                                        [default: local]                  │</span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>    <span class="at">--help</span>       <span class="at">-h</span>                     Show this message and exit.       │</span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a><span class="ex">╰──────────────────────────────────────────────────────────────────────────╯</span></span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true" tabindex="-1"></a><span class="ex">╭─</span> Method parameters ──────────────────────────────────────────────────────╮</span>
<span id="cb4-24"><a href="#cb4-24" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span> <span class="at">--fragsize</span>        INTEGER RANGE [x<span class="op">&gt;</span>=1]  Comparison method fragment size. │</span>
<span id="cb4-25"><a href="#cb4-25" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>                                         [default: 1020]                  │</span>
<span id="cb4-26"><a href="#cb4-26" aria-hidden="true" tabindex="-1"></a><span class="ex">╰──────────────────────────────────────────────────────────────────────────╯</span></span>
<span id="cb4-27"><a href="#cb4-27" aria-hidden="true" tabindex="-1"></a><span class="ex">╭─</span> Debugging ──────────────────────────────────────────────────────────────╮</span>
<span id="cb4-28"><a href="#cb4-28" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span> <span class="at">--temp</span>         DIRECTORY  Directory to use for intermediate files, which │</span>
<span id="cb4-29"><a href="#cb4-29" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>                           for debugging purposes will not be deleted.    │</span>
<span id="cb4-30"><a href="#cb4-30" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>                           For clusters this must be on a shared drive.   │</span>
<span id="cb4-31"><a href="#cb4-31" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>                           Default behaviour is to use a system specified │</span>
<span id="cb4-32"><a href="#cb4-32" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>                           temporary directory <span class="er">(</span><span class="ex">specific</span> to the           │</span>
<span id="cb4-33"><a href="#cb4-33" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>                           compute-node when using a cluster<span class="kw">)</span> <span class="ex">and</span> remove  │</span>
<span id="cb4-34"><a href="#cb4-34" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>                           this afterwards.                               │</span>
<span id="cb4-35"><a href="#cb4-35" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span> <span class="at">--wtemp</span>        DIRECTORY  Directory to use for temporary workflow        │</span>
<span id="cb4-36"><a href="#cb4-36" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>                           coordination files, which for debugging        │</span>
<span id="cb4-37"><a href="#cb4-37" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>                           purposes will not be deleted. For clusters     │</span>
<span id="cb4-38"><a href="#cb4-38" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>                           this must be on a shared drive. Default        │</span>
<span id="cb4-39"><a href="#cb4-39" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>                           behaviour is to use a system specified         │</span>
<span id="cb4-40"><a href="#cb4-40" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>                           temporary directory <span class="er">(</span><span class="cf">for</span> the <span class="bu">local</span> <span class="va">executor</span><span class="kw">)</span>   <span class="ex">│</span></span>
<span id="cb4-41"><a href="#cb4-41" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>                           or a temporary directory under the present     │</span>
<span id="cb4-42"><a href="#cb4-42" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>                           direct <span class="er">(</span><span class="cf">for</span> clusters<span class="kw">)</span><span class="ex">,</span> and remove this         │</span>
<span id="cb4-43"><a href="#cb4-43" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>                           afterwards.                                    │</span>
<span id="cb4-44"><a href="#cb4-44" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span> <span class="at">--log</span>          FILE       Where to record log<span class="er">(</span><span class="ex">s</span><span class="kw">)</span><span class="bu">.</span> Use <span class="st">'-'</span> for no         │</span>
<span id="cb4-45"><a href="#cb4-45" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>                           logging.                                       │</span>
<span id="cb4-46"><a href="#cb4-46" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>                           [default: pyani-plus.log]                      │</span>
<span id="cb4-47"><a href="#cb4-47" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span> <span class="at">--debug</span>                   Show debugging level logging at the terminal   │</span>
<span id="cb4-48"><a href="#cb4-48" aria-hidden="true" tabindex="-1"></a><span class="ex">│</span>                           <span class="er">(in</span> <span class="ex">addition</span> to the log file<span class="kw">)</span><span class="bu">.</span>                 │</span>
<span id="cb4-49"><a href="#cb4-49" aria-hidden="true" tabindex="-1"></a><span class="ex">╰──────────────────────────────────────────────────────────────────────────╯</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<section id="collect-genomes-for-analysis" class="level2">
<h2 class="anchored" data-anchor-id="collect-genomes-for-analysis">Collect genomes for analysis</h2>
<p>While you can work with your choice of genomes as FASTA files placed in a local directory, we suggest using the ten genomes provided when first following this walkthrough to ensure the output matches the expected results.</p>
<p><code>pyANI-plus</code> accepts FASTA files with the extensions <code>.fasta</code>, <code>.fas</code>, and <code>.fna</code>, along with gzip compressed versions like <code>.fasta.gz</code>, <code>.fas.gz</code>, and <code>.fna.gz</code>. Please make sure that your input files match these extensions to ensure that <code>pyANI-plus</code> works.</p>
</section>
<section id="conducting-ani-analysis" class="level2">
<h2 class="anchored" data-anchor-id="conducting-ani-analysis">Conducting ANI analysis</h2>
<p><code>pyANI-plus</code> enables genome comparison using various ANI methods. In this walkthrough, we will demonstrate methods such as ANIm, ANIb, dnadiff, FastANI, and Sourmash. While running all methods is not mandatory, we recommend doing so, as we will later compare their results using the <code>plot-run-comp</code> command.</p>
<p>Running any ANI method on requires you specify the directory containing the genome data (e.g., <code>walkthrough_data/</code>), and the path to the pyANI-plus SQLite3 database (<code>walkthrough.db</code> for this walkthrough).</p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<p>If this is your first analysis and the SQLite3 database does not yet exist, you must use the <code>--create-db</code> option; otherwise, you’ll encounter the following error:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="ex">ERROR:</span> Database walkthrough.db does not exist, but not using <span class="at">--create-db</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<p>Optionally, you can provide a custom name for the analysis with the <code>--name</code> option for easier reference.</p>
<p>If you want to run the ANI analysis on a cluster using SLURM, you must set the execution method using <code>--executor slurm</code> (the default is <code>local</code>, see later).</p>
<section id="conduct-anib-analysis" class="level3">
<h3 class="anchored" data-anchor-id="conduct-anib-analysis">Conduct ANIb analysis</h3>
<p>In this walkthrough, we will first run the ANIb analysis on the downloaded genomes using the following command line:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="ex">pyani-plus</span> anib walkthrough_data/ <span class="at">--database</span> walkthrough.db <span class="at">--create-db</span> <span class="at">--name</span> <span class="st">"walkthrough ANIb"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>With the suggested ten genomes this takes about 5 minutes on a recent multi-core desktop computer. The other methods should all be faster.</p>
<p>If you wish you can select a different fragment size for the comparison method using the <code>--fragsize</code> option. The default size is 1020bp (listed in the help output shown earlier), which is typically used for ANIb.</p>
</section>
<section id="conduct-anim-analysis" class="level3">
<h3 class="anchored" data-anchor-id="conduct-anim-analysis">Conduct ANIm analysis</h3>
<p>Next, we will run the ANIm analysis on the same genomes using the following command line:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="ex">pyani-plus</span> anim walkthrough_data/ <span class="at">--database</span> walkthrough.db <span class="at">--name</span> <span class="st">"walkthrough ANIm"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>In ANIm analysis, the default setting uses anchor matches that are unique in the reference but not necessarily unique in the query (<code>--mode mum</code>). You can change this to include all anchor matches, regardless of their uniqueness, by setting the <code>--mode</code> option to <code>maxmatch</code>.</p>
</section>
<section id="conduct-dnadiff-analysis" class="level3">
<h3 class="anchored" data-anchor-id="conduct-dnadiff-analysis">Conduct dnadiff analysis</h3>
<p>To compare genomes in the input <code>walkthrough_data</code> directory using <code>dnadiff</code> method use the following command line:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="ex">pyani-plus</span> dnadiff walkthrough_data/ <span class="at">--database</span> walkthrough.db <span class="at">--name</span> <span class="st">"walkthrough dnadiff"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="conduct-fastani-analysis" class="level3">
<h3 class="anchored" data-anchor-id="conduct-fastani-analysis">Conduct fastani analysis</h3>
<p>To run fastani analysis on the genomes in the input <code>walkthrough_data</code> directory use the following command line:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="ex">pyani-plus</span> fastani walkthrough_data/ <span class="at">--database</span> walkthrough.db <span class="at">--name</span> <span class="st">"walkthrough fastani"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>In <code>fastani</code> analysis, additional method parameters can be changed by the user. These include:</p>
<ul>
<li><code>--fragsize</code>: Fragment length used in the analysis (default: <code>3000</code>).</li>
<li><code>--kmersize</code>: K-mer size, set to <code>16</code> by default. It can be set to any value smaller than 16.</li>
<li><code>--minmatch</code>: Minimum fraction of the genome that must be shared for ANI to be considered reliable. If the reference and query genome sizes differ, the smaller genome is used. (Default: <code>0.2</code>).</li>
</ul>
</section>
<section id="conduct-sourmash-analysis" class="level3">
<h3 class="anchored" data-anchor-id="conduct-sourmash-analysis">Conduct sourmash analysis</h3>
<p>Lastly, we can run sourmash analysis with the following command line:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="ex">pyani-plus</span> sourmash walkthrough_data/ <span class="at">--database</span> walkthrough.db <span class="at">--name</span> <span class="st">"walkthrough sourmash"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>For <code>sourmash</code> analysis, additional method parameters can be changed by the user. These include: - <code>--scaled</code>: Compression ration (defult: <code>1000</code>) - <code>--kmersize</code>: K-mer size (default: <code>31</code>)</p>
</section>
</section>
<section id="reporting-analyses-and-analysis-results" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="reporting-analyses-and-analysis-results">Reporting Analyses and Analysis Results</h2>
<section id="list-runs" class="level3">
<h3 class="anchored" data-anchor-id="list-runs">List all runs in the database</h3>
<p><code>pyANI-plus</code> enables you to view all runs defined in a SQLite3 database. To display all runs from the database (eg. <code>walkthrough.db</code> for this walkthrough), use this command:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="ex">pyani-plus</span> list-runs <span class="at">--database</span> walkthrough.db</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>You will see the following table, or something similar, depending on the analyses contained within the database, displayed on your screen:</p>
<pre><code>                              5 analysis runs in walkthrough.db                              
┏━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━┳━━━━━━┳━━━━━━┳━━━━━━━━━┳━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━┓
┃ ID ┃ Date       ┃ Method   ┃ Done ┃ Null ┃ Miss ┃ Total   ┃ Status ┃ Name                 ┃
┡━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━╇━━━━━━╇━━━━━━╇━━━━━━━━━╇━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━┩
│  1 │ 2025-03-17 │ ANIb     │  100 │    0 │    0 │ 100=10² │ Done   │ walkthrough ANIb     │
│  2 │ 2025-03-17 │ ANIm     │   68 │   32 │    0 │ 100=10² │ Done   │ walkthrough ANIm     │
│  3 │ 2025-03-17 │ dnadiff  │  100 │    0 │    0 │ 100=10² │ Done   │ walkthrough dnadiff  │
│  4 │ 2025-03-17 │ fastANI  │   68 │   32 │    0 │ 100=10² │ Done   │ walkthrough fastani  │
│  5 │ 2025-03-17 │ sourmash │   68 │   32 │    0 │ 100=10² │ Done   │ walkthrough sourmash │
└────┴────────────┴──────────┴──────┴──────┴──────┴─────────┴────────┴──────────────────────┘</code></pre>
<p>In this table, each row represents a single run, and the columns provide the following details:</p>
<ul>
<li><code>ID</code>: Unique ID for the run</li>
<li><code>Date</code>: Date when the analysis was executed</li>
<li><code>Method</code>: ANI method used</li>
<li><code>Done</code>: Number of completed ANI comparisons</li>
<li><code>Null</code>: Number of analyses where no alignment was found (e.g., comparisons between highly divergent genomes)</li>
<li><code>Miss</code>: Number of comparisons that were not completed</li>
<li><code>Status</code>: Current status of the analysis—e.g., Done for completed analyses, Running for comparisons that have started but are still in progress</li>
<li><code>Name</code>: Run name.</li>
</ul>
</section>
<section id="exporting-ani-results-in-a-tabular-format" class="level3">
<h3 class="anchored" data-anchor-id="exporting-ani-results-in-a-tabular-format">Exporting ANI results in a tabular format</h3>
<p><code>pyANI-plus</code> allows ANI results to be exported in a tabular format, but the output directory must already be present. In this example, we create a <code>walkthrough_output/</code> directory and use the following command to export results for the ANIb analysis:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mkdir</span> output <span class="co"># create directory called output</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="ex">pyani-plus</span> export-run <span class="at">--database</span> walkthrough.db <span class="at">--outdir</span> walkthrough_output/ <span class="at">--run-id</span> 1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>If <code>--run-id</code> is not specified the latest run will be exported. To export runs for other analyses, you can specify the <code>--run-id</code> by matching the ID number provided in the table provided by <a href="./walkthrough.html#list-runs">list-runs subcommand</a>.</p>
</div>
</div>
<p>This will report the relevant information to new files in the <code>walkthrough_output/</code> directory. The matrix output files are named <code>&lt;method&gt;_&lt;property&gt;.tsv</code> while the long form is named <code>&lt;method&gt;_run_&lt;run-id&gt;.tsv</code> and will include the query and subject genomes and all the comparison properties as columns:</p>
<pre><code>.
├── ANIb_aln_lengths.tsv
├── ANIb_hadamard.tsv
├── ANIb_identity.tsv
├── ANIb_query_cov.tsv
├── ANIb_run_1.tsv
├── ANIb_sim_errors.tsv
└── ANIb_tANI.tsv</code></pre>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<p>Incomplete runs will return an error. There will be no output for empty run. For partial runs the long form table will be exported, but not the matrices.</p>
</div>
</div>
</section>
<section id="graphical-output" class="level3 page-columns page-full">
<h3 class="anchored" data-anchor-id="graphical-output">Graphical output</h3>
<p>Graphical output (<code>JPG</code>, <code>PDF</code>, <code>PNG</code> and <code>SVGZ</code> formats) is obtanined by executing the <code>pyani-plus plot-run</code> subcommand, specifying the database and ouput directory:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="ex">pyani-plus</span> plot-run <span class="at">--database</span> walkthrough.db <span class="at">--outdir</span> walkthrough_output/ <span class="at">--run-id</span> 2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>If <code>--run-id</code> is not specified the latest run will be used to generate graphical output. To generate graphical output for other analyses, you can specify the <code>--run-id</code> by matching the ID number provided in the table provided by <a href="./walkthrough.html#list-runs">list-runs subcommand</a>.</p>
<p>Optionally, you can label genomes using <code>md5</code> and <code>filename</code> using the <code>--label</code> option.</p>
</div>
</div>
<p><code>plot-run</code> subcommand generates the following heatmaps, distribution plots, and tabular output from an ANI analysis for a specified <code>--run_id</code>, using data stored in a local SQLite3 database:</p>
<ul>
<li>percentage identity of aligned regions (<code>&lt;method&gt;_idenity_heatmap.&lt;extension&gt;</code>, <code>&lt;method&gt;_identity_dist.&lt;extension&gt;</code> and <code>&lt;method&gt;_identity.tsv</code>, <a href="#fig-anim-heatmap" class="quarto-xref">Figure&nbsp;<span>1</span></a>)</li>
</ul>

<div class="no-row-height column-margin column-container"><div class="">
<div id="fig-anim-heatmap" class="lightbox quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-anim-heatmap-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<a href="walkthrough_plots/ANIm_identity_heatmap.png" class="lightbox" data-gallery="quarto-lightbox-gallery-1" title="Figure&nbsp;1: ANIm percentage identity heatmap for walkthrough_data"><img src="walkthrough_plots/ANIm_identity_heatmap.png" class="img-fluid figure-img"></a>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-anim-heatmap-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;1: ANIm percentage identity heatmap for walkthrough_data
</figcaption>
</figure>
</div>
</div></div><p>In the above heatmap, each cell represents a pairwise comparision between the genomes shown in the rows and columns, showing the pairwise identity of aligned regions. The dendrograms are single-linkage clustering trees generated from the matrix of pairwise identity results. The default color scheme assigns red to cells with identity <span class="math inline">\(\geq 0.95\)</span>, blue to those with identity <span class="math inline">\(&lt;0.95\)</span>, and orange to cells representing comparisons with no alignment found (e.g., NULLs). This division corresponds to a widely-used convention for bacterial species boundaries.</p>
<ul>
<li>percentage coverage of each query genome by aligned regions (<code>&lt;method&gt;_query_cov_heatmap.&lt;extension&gt;</code>, <code>&lt;method&gt;_query_cov_dist.&lt;extension&gt;</code>, <code>&lt;method&gt;_query_cov_scatter.&lt;extension&gt;</code> and <code>&lt;method&gt;_query_cov.tsv</code>)</li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="walkthrough_plots/ANIm_query_cov_heatmap.png" class="img-fluid figure-img"></p>
<figcaption>ANIm percentage coverage heatmap for walkthrough_data</figcaption>
</figure>
</div>
<p>In the above heatmap, each cell represents a pairwise comparision between the genomes shown in the rows and columns, showing the pairwise coverage of each genome by aligned regions in the comparison. The dendrograms are single-linkage clustering trees generated from the matrix of pairwise coverage results. The default color scheme assigns red to cells with coverage <span class="math inline">\(\geq 0.50\)</span>, blue to those with coverage <span class="math inline">\(&lt;0.50\)</span>, and orange to cells representing comparisons with no alignment found (e.g., NULLs). This division corresponds to a strict majority of each genome in the comparison being alignable (a plausible minimum requirement for two sequences being considered “the same thing”).</p>
<ul>
<li>a Hadamard matrix of percentage identity multiplied by percentage coverage for each comparison (<code>&lt;method&gt;_hadamard_heatmap.&lt;extension&gt;</code>, <code>&lt;method&gt;_hadamard_dist.&lt;extension&gt;</code> and <code>&lt;method&gt;_hadamard.tsv</code>)</li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="walkthrough_plots/ANIm_hadamard_heatmap.png" class="img-fluid figure-img"></p>
<figcaption>ANIm hadamard heatmap for walkthrough_data</figcaption>
</figure>
</div>
<ul>
<li>a total Avenarge Nucleotide Identity (tANI) matrix of he negative log of the coverage multiplied by identity (<code>&lt;method&gt;_tANI_heatmap.&lt;extension&gt;</code>, <code>&lt;method&gt;_tANI_dist.&lt;extension&gt;</code>, <code>&lt;method&gt;_tANI_scatter.&lt;extension&gt;</code> and <code>&lt;method&gt;_tANI.tsv</code>)</li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="walkthrough_plots/ANIm_tANI_heatmap.png" class="img-fluid figure-img"></p>
<figcaption>ANIm tANI heatmap for walkthrough_data</figcaption>
</figure>
</div>
<ul>
<li>number of “similarity errors” on each genome (<code>&lt;method&gt;_sim_errors.tsv</code>)</li>
<li>Long form of ANI results which include the query and subject genomes and all the comparison properties as columns (<code>&lt;method&gt;_run_&lt;ID&gt;.tsv</code>)</li>
</ul>
</section>
</section>
<section id="plotting-comparisons-between-runs" class="level2">
<h2 class="anchored" data-anchor-id="plotting-comparisons-between-runs">Plotting comparisons between runs</h2>
<p>ANI results can vary depending on the method used. <code>pyANI-plus</code> allows you to compare the ANI results from multiple runs. In this example, we show how to use the <code>pyani-plus plot-run-comp</code> subcommand to visualise and compare these results. Running <code>pyani-plus plot-run-comp</code> requires specifying the output directory (e.g., <code>walkthrough_output/</code>), the path to the pyANI-plus SQLite3 database (<code>walkthrough.db</code> for this walkthrough), and a comma-separated list of run IDs for comparison.</p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<p>The first run ID will be treated as the reference, and all subsequent runs will be compared to it.</p>
<p>In this example, we use ANIb as the reference method, with other methods compared against it.</p>
</div>
</div>
<div class="sourceCode" id="cb16"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="ex">pyani-plus</span> plot-run-comp <span class="at">--database</span> walkthrough.db <span class="at">--outdir</span> walkthrough_output/ <span class="at">--run-ids</span> 1,2,3,4,5</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>This command generates the following outputs:</p>
<ul>
<li>A set of scatterplots where the X-axis represents genome identity from the reference method (here, ANIb), and the Y-axis represents genome identity from the compared methods/runs (<code>&lt;reference_method&gt;_identity_&lt;run_ID&gt;_scatter_vs_others.&lt;extension&gt;</code>).</li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="walkthrough_plots/ANIb_identity_1_scatter_vs_others.png" class="img-fluid figure-img"></p>
<figcaption>Scatterplot output for <code>pyani-plus plot-run-comp</code> subcommand</figcaption>
</figure>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>The red diagonal line serves as a reference, indicating where data points should align on the scatterplot if the pairwise comparison results match between methods.</p>
</div>
</div>
<ul>
<li>A set of scatterplots showing absolute differences between pairwise comparisons, with the X-axis representing genome identity from the reference method (here, ANIb) and the Y-axis showing the difference in genome identity for the compared methods/runs.</li>
</ul>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="walkthrough_plots/ANIb_identity_1_diff_vs_others.png" class="img-fluid figure-img"></p>
<figcaption>Scatterplot output for <code>pyani-plus plot-run-comp</code> subcommand</figcaption>
</figure>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>The red horizontal line at 0 indicates no difference between the pairwise comparison results across methods.</p>
</div>
</div>
<ul>
<li>Tabular summary of the comparison output (<code>&lt;reference_method&gt;_identity_&lt;run_ID&gt;_vs_&lt;run_ID&gt;.tsv</code>).</li>
</ul>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./installation.html" class="pagination-link" aria-label="Installation">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Installation</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./subcommands/subcommands.html" class="pagination-link" aria-label="pyANI-plus subcommands">
        <span class="nav-page-text">pyANI-plus subcommands</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p>pyANI-plus documentation</p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    <div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/sipbs-compbiol/REPOSITORY/edit/main/walkthrough.qmd" class="toc-action"><i class="bi bi-github"></i>Edit this page</a></li><li><a href="https://github.com/sipbs-compbiol/REPOSITORY/issues/new" class="toc-action"><i class="bi empty"></i>Report an issue</a></li></ul></div></div>
    <div class="nav-footer-right">
<p>This book was built with <a href="https://quarto.org/">Quarto</a>.</p>
</div>
  </div>
</footer>
<script>var lightboxQuarto = GLightbox({"closeEffect":"zoom","descPosition":"bottom","loop":false,"openEffect":"zoom","selector":".lightbox"});
(function() {
  let previousOnload = window.onload;
  window.onload = () => {
    if (previousOnload) {
      previousOnload();
    }
    lightboxQuarto.on('slide_before_load', (data) => {
      const { slideIndex, slideNode, slideConfig, player, trigger } = data;
      const href = trigger.getAttribute('href');
      if (href !== null) {
        const imgEl = window.document.querySelector(`a[href="${href}"] img`);
        if (imgEl !== null) {
          const srcAttr = imgEl.getAttribute("src");
          if (srcAttr && srcAttr.startsWith("data:")) {
            slideConfig.href = srcAttr;
          }
        }
      } 
    });
  
    lightboxQuarto.on('slide_after_load', (data) => {
      const { slideIndex, slideNode, slideConfig, player, trigger } = data;
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(slideNode);
      }
    });
  
  };
  
})();
          </script>




</body></html>