# `anib` {#anib-subcommand .unnambered}

## Usage
The `anib` subcommand processes genome files from the `indir` directory to perform ANIb analysis, and logs comparison and run data in a local SQLite3 database.

```shell
$ pyani-plus anib -h
                                                                                     
 Usage: pyani-plus anib [OPTIONS] FASTA                                         
                                                                                
 Execute ANIb calculations, logged to a pyANI-plus SQLite3 database.            
                                                                                
╭─ Arguments ──────────────────────────────────────────────────────────────────╮
│ *    fasta      PATH  Directory of FASTA files (extensions .fas, .fasta,     │
│                       .fna).                                                 │
│                       [required]                                             │
╰──────────────────────────────────────────────────────────────────────────────╯
╭─ Options ────────────────────────────────────────────────────────────────────╮
│ *  --database   -d      FILE           Path to pyANI-plus SQLite3 database.  │
│                                        [required]                            │
│    --name               TEXT           Run name. Default is 'N genomes using │
│                                        METHOD'.                              │
│    --create-db                         Create database if does not exist.    │
│    --executor           [local|slurm]  How should the internal tools be run? │
│                                        [default: local]                      │
│    --help       -h                     Show this message and exit.           │
╰──────────────────────────────────────────────────────────────────────────────╯
╭─ Method parameters ──────────────────────────────────────────────────────────╮
│ --fragsize        INTEGER RANGE [x>=1]  Comparison method fragment size.     │
│                                         [default: 1020]                      │
╰──────────────────────────────────────────────────────────────────────────────╯
╭─ Debugging ──────────────────────────────────────────────────────────────────╮
│ --temp         DIRECTORY  Directory to use for intermediate files, which for │
│                           debugging purposes will not be deleted. For        │
│                           clusters this must be on a shared drive. Default   │
│                           behaviour is to use a system specified temporary   │
│                           directory (specific to the compute-node when using │
│                           a cluster) and remove this afterwards.             │
│ --wtemp        DIRECTORY  Directory to use for temporary workflow            │
│                           coordination files, which for debugging purposes   │
│                           will not be deleted. For clusters this must be on  │
│                           a shared drive. Default behaviour is to use a      │
│                           system specified temporary directory (for the      │
│                           local executor) or a temporary directory under the │
│                           present direct (for clusters), and remove this     │
│                           afterwards.                                        │
│ --log          FILE       Where to record log(s). Use '-' for no logging.    │
│                           [default: pyani-plus.log]                          │
╰──────────────────────────────────────────────────────────────────────────────╯

```

## What does `aniB` method do?
`pyani-plus` implements Average Nucleotide Identity using [BLASTN](https://blast.ncbi.nlm.nih.gov/doc/blast-help/downloadblastdata.html#downloadblastdata) (ANIb) as described in @goris2007dna.


In brief, the analysis proceeds as follows for a set of input prokaryotic genomes:

- __Genome Fragmentation__: Query genomes are fragmented into 1020bp segments. Fragments shorter than 1020 bp are retained.

- __Sequence Alignment__: The fragmented query sequences are aligned against the intact subject sequences using [BLASTN](https://blast.ncbi.nlm.nih.gov/doc/blast-help/downloadblastdata.html#downloadblastdata).
- __Result Parsing and Calculation__: The [BLASTN](https://blast.ncbi.nlm.nih.gov/doc/blast-help/downloadblastdata.html#downloadblastdata) output (tab-separated plain text format) is parsed, and the following metrics are computed:

    - __Aligned bases__: Total aligned bases for each genome (`alignment_length - gaps`)

    - __Genome coverage__: The fraction of each genome (query and subject) that is aligned (`alignment_length / query_genome_length`)

    - __Average Nucleotide Identity (ANI)__: The proportion of aligned regions that are identical

    - __Similarity errors__: The number of unaligned or non-identical bases (`gaps + mismatches`)

    - __Hadamard score__: The percentage identity multiplied by percentage coverage (`ANI * coverage`)

    - __tANI score__: The negative natural log of coverage multiplied by identity

- __Recording values__: The output values are recorded in the [SQLite3](https://www.sqlite.org/) database.

{{< include ../includes/_bidirectional_comp.qmd >}}

{{< include ../includes/_method_output.qmd >}}

## References {.unnumbered}
