# `external-alignment` {#external-alignment-subcommand .unnambered}

The `external-alignment` subcommand compute pairwise ANI from given multiple-sequence-alignment (MSA) file and genome files from the `indir` directory, and logs comparison and run data in a local SQLite3 database.


```shell
$ pyani-plus external-alignment -h
                                                                                     
 Usage: pyani-plus external-alignment [OPTIONS] FASTA                                
                                                                                     
 Compute pairwise ANI from given multiple-sequence-alignment (MSA) file.             
                                                                                     
╭─ Arguments ───────────────────────────────────────────────────────────────────────╮
│ *    fasta      PATH  Directory of FASTA files (extensions .fas, .fasta, .fna)    │
│                       [required]                                                  │
╰───────────────────────────────────────────────────────────────────────────────────╯
╭─ Options ─────────────────────────────────────────────────────────────────────────╮
│ *  --database           FILE           Path to pyANI-plus SQLite3 database        │
│                                        [required]                                 │
│    --name               TEXT           Run name. Default is 'N genomes using      │
│                                        METHOD'.                                   │
│    --create-db                         Create database if does not exist          │
│    --executor           [local|slurm]  How should the internal tools be run?      │
│                                        [default: local]                           │
│    --help       -h                     Show this message and exit.                │
╰───────────────────────────────────────────────────────────────────────────────────╯
╭─ Debugging ───────────────────────────────────────────────────────────────────────╮
│ --temp         DIRECTORY  Directory to use for intermediate files, which for      │
│                           debugging purposes will not be deleted. For clusters    │
│                           this must be on a shared drive. Default behaviour is to │
│                           use a system specified temporary directory (specific to │
│                           the compute-node when using a cluster) and remove this  │
│                           afterwards.                                             │
│ --wtemp        DIRECTORY  Directory to use for temporary workflow coordination    │
│                           files, which for debugging purposes will not be         │
│                           deleted. For clusters this must be on a shared drive.   │
│                           Default behaviour is to use a system specified          │
│                           temporary directory (for the local executor) or a       │
│                           temporary directory under the present direct (for       │
│                           clusters), and remove this afterwards.                  │
╰───────────────────────────────────────────────────────────────────────────────────╯
╭─ Method parameters ───────────────────────────────────────────────────────────────╮
│ *  --alignment        FILE                 FASTA format MSA of the same genomes   │
│                                            (one sequence per genome)              │
│                                            [default: None]                        │
│                                            [required]                             │
│    --label            [md5|filename|stem]  How are the sequences in the MSA       │
│                                            labelled vs the FASTA genomes?         │
│                                            [default: stem]                        │
╰───────────────────────────────────────────────────────────────────────────────────╯
```