The basic structure of a command in `pyANI-plus` follows this general format:

```shell
pyani-plus [OPTIONS] COMMAND [ARGS]
```

To view the avaliable options and commands for `pyANI-plus`, use the `-h` (help) option. For example:

```shell
$ pyani-plus -h
 Usage: pyani-plus [OPTIONS] COMMAND [ARGS]...                           
                                                                         
╭─ Options ─────────────────────────────────────────────────────────────╮
│ --install-completion            Install completion for the current    │
│                                 shell.                                │
│ --show-completion               Show completion for the current       │
│                                 shell, to copy it or customize the    │
│                                 installation.                         │
│ --help                -h        Show this message and exit.           │
╰───────────────────────────────────────────────────────────────────────╯
╭─ Commands ────────────────────────────────────────────────────────────╮
│ classify             Classify genomes into clusters based on ANI      │
│                      results.                                         │
│ delete-run           Delete any single run from the given pyANI-plus  │
│                      SQLite3 database.                                │
│ export-run           Export any single run from the given pyANI-plus  │
│                      SQLite3 database.                                │
│ list-runs            List the runs defined in a given pyANI-plus      │
│                      SQLite3 database.                                │
│ plot-run             Plot heatmaps and distributions for any single   │
│                      run.                                             │
│ plot-run-comp        Plot comparisons between multiple runs.          │
│ resume               Resume any (partial) run already logged in the   │
│                      database.                                        │
╰───────────────────────────────────────────────────────────────────────╯
╭─ ANI methods ─────────────────────────────────────────────────────────╮
│ anib                 Execute ANIb calculations, logged to a           │
│                      pyANI-plus SQLite3 database.                     │
│ anim                 Execute ANIm calculations, logged to a           │
│                      pyANI-plus SQLite3 database.                     │
│ dnadiff              Execute mumer-based dnadiff calculations, logged │
│                      to a pyANI-plus SQLite3 database.                │
│ external-alignment   Compute pairwise ANI from given                  │
│                      multiple-sequence-alignment (MSA) file.          │
│ fastani              Execute fastANI calculations, logged to a        │
│                      pyANI-plus SQLite3 database.                     │
│ sourmash             Execute sourmash-plugin-branchwater ANI          │
│                      calculations, logged to a pyANI-plus SQLite3     │
│                      database.                                        │
╰───────────────────────────────────────────────────────────────────────╯

```

To see the options for a specific subcommand, use `pyani-plus <subcommand> -h`. For example, to view options for the ANIb method:

```shell
$ pyani-plus anib -h

 Usage: pyani-plus anib [OPTIONS] FASTA                                         
                                                                                
 Execute ANIb calculations, logged to a pyANI-plus SQLite3 database.            
                                                                                
╭─ Arguments ──────────────────────────────────────────────────────────────────╮
│ *    fasta      PATH  Directory of FASTA files (extensions .fas, .fasta,     │
│                       .fna).                                                 │
│                       [required]                                             │
╰──────────────────────────────────────────────────────────────────────────────╯
╭─ Options ────────────────────────────────────────────────────────────────────╮
│ *  --database   -d      FILE           Path to pyANI-plus SQLite3 database.  │
│                                        [required]                            │
│    --name               TEXT           Run name. Default is 'N genomes using │
│                                        METHOD'.                              │
│    --create-db                         Create database if does not exist.    │
│    --executor           [local|slurm]  How should the internal tools be run? │
│                                        [default: local]                      │
│    --help       -h                     Show this message and exit.           │
╰──────────────────────────────────────────────────────────────────────────────╯
╭─ Method parameters ──────────────────────────────────────────────────────────╮
│ --fragsize        INTEGER RANGE [x>=1]  Comparison method fragment size.     │
│                                         [default: 1020]                      │
╰──────────────────────────────────────────────────────────────────────────────╯
╭─ Debugging ──────────────────────────────────────────────────────────────────╮
│ --temp         DIRECTORY  Directory to use for intermediate files, which for │
│                           debugging purposes will not be deleted. For        │
│                           clusters this must be on a shared drive. Default   │
│                           behaviour is to use a system specified temporary   │
│                           directory (specific to the compute-node when using │
│                           a cluster) and remove this afterwards.             │
│ --wtemp        DIRECTORY  Directory to use for temporary workflow            │
│                           coordination files, which for debugging purposes   │
│                           will not be deleted. For clusters this must be on  │
│                           a shared drive. Default behaviour is to use a      │
│                           system specified temporary directory (for the      │
│                           local executor) or a temporary directory under the │
│                           present direct (for clusters), and remove this     │
│                           afterwards.                                        │
│ --log          FILE       Where to record log(s). Use '-' for no logging.    │
│                           [default: pyani-plus.log]                          │
╰──────────────────────────────────────────────────────────────────────────────╯
```